src/lib/dnssd/IncrementalResolve.cpp

```diff
@@ -0,0 +1,374 @@
+        return ServiceNameType::kOperational;
+    }
+    else if (name == kCommissionableSuffix)
+    {
+        return ServiceNameType::kCommissionable;
+    }
+    else if (name == kCommissionerSuffix)
+    {
+        return ServiceNameType::kCommissioner;
+    }
+    else
+    {
+        return ServiceNameType::kInvalid;
+    }
+}
+
+/// Checks if the name is of the form <something>._sub._matter(c|d)._udp.local
+bool IsCommissionSubtype(SerializedQNameIterator name)
+{
+    if (!name.Next() || !name.IsValid())
+    {
+        // subtype should be a prefix
+        return false;
+    }
+
+    return (name == kCommissionerSubTypeSuffix) || (name == kCommissionableSubTypeSuffix);
+}
+
+} // namespace
+
+CHIP_ERROR StoredServerName::Set(SerializedQNameIterator value)
+{
+    chip::Encoding::BigEndian::BufferWriter output(mBuffer, sizeof(mBuffer));
+    RecordWriter writer(&output);
+
+    writer.WriteQName(value);
+
+    if (!writer.Fit())
+    {
+        Clear();
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+SerializedQNameIterator StoredServerName::Get() const
+{
+    return SerializedQNameIterator(BytesRange(mBuffer, mBuffer + sizeof(mBuffer)), mBuffer);
+}
+
+CHIP_ERROR IncrementalResolver::InitializeParsing(mdns::Minimal::SerializedQNameIterator name, const mdns::Minimal::SrvRecord & srv)
+{
+    mSpecificResolutionData = SpecificParseData();
+    mCommonResolutionData.Reset();
+
+    ReturnErrorOnFailure(mRecordName.Set(name));
+    ReturnErrorOnFailure(mServerName.Set(srv.GetName()));
+    mCommonResolutionData.port = srv.GetPort();
+
+    {
+        // TODO: Chip code historically seems to assume that the host name is of the
+        // form "<MAC or 802.15.4 Extended Address in hex>.local" and only saves the first part.
+        //
+        // This should not be needed as server name should not be relevant once parsed.
+        SerializedQNameIterator serverName = srv.GetName();
+
+        VerifyOrReturnError(serverName.Next() && serverName.IsValid(), CHIP_ERROR_INVALID_ARGUMENT);
+
+        // only save the first part: the MAC or 802.15.4 Extended Address in hex
+        Platform::CopyString(mCommonResolutionData.hostName, serverName.Value());
+    }
+
+    switch (ComputeServiceNameType(name))
+    {
+    case ServiceNameType::kOperational:
+        mSpecificResolutionData.Set<OperationalNodeData>();
+        {
+            // Operational addresses start with peer node information
+            SerializedQNameIterator nameCopy = name;
+            if (!nameCopy.Next() || !nameCopy.IsValid())
+            {
+                mSpecificResolutionData = SpecificParseData();
+                return CHIP_ERROR_INVALID_ARGUMENT;
+            }
+
+            CHIP_ERROR err =
+                ExtractIdFromInstanceName(nameCopy.Value(), &mSpecificResolutionData.Get<OperationalNodeData>().peerId);
+            if (err != CHIP_NO_ERROR)
+            {
+                mSpecificResolutionData = SpecificParseData();
```

Comment:
If you do this, could the request eventually await more records forever if the parsing errors and you set this empty?

Reply:
Then maybe have a `Reset()`/`ResetToInactive()` method, which avoids the mental burden of figuring-out your internal conventions.

Reply:
Added method and switched to RAII here.